### PR TITLE
Add an option to AnnounceOnBuild/AnnounceOnKill to only announce their activity to it's owner.

### PR DESCRIPTION
**Pull request by GraionDilach**
Requested by @ABrandau and SchismNavigator.

### CODE REVIEW

**Comment by abcdefg30** [comment]
Um, I know you'll probably hate me for that scope creep, but couldn't we get some kind of stances flag instead? Like have an option for `only the owner`, `allies`(, `neutrals`) and `enemies`.

**Comment by GraionDilach** [comment]
Yes. I will. I haven't started looking into the stances setup and usecases as part of the #10892 fix either. But if that's what everyone wants, then I'll look into it (honestly, I wanted to mod today and only did this as a 5 minute break).

File: OpenRA.Mods.Common/Traits/Sound/AnnounceOnBuild.cs [hunk 1]
```diff
@@ -19,6 +19,9 @@ public class AnnounceOnBuildInfo : ITraitInfo
 		[Desc("Voice to use when built/trained.")]
 		[VoiceReference] public readonly string Voice = "Build";
 
+		[Desc("Should the voice be played for the owner alone?")]
+		public readonly bool AnnounceOnlyToOwner = false;
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Sound/AnnounceOnBuild.cs]: `OnlyToOwner` instead maybe? It is otherwise a bit redundant.

Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/Sound/AnnounceOnBuild.cs]: Considering we have some English-pickers in the form of @Phrohdoh I thought it's better to have it redundant than nonEnglish.

Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Sound/AnnounceOnBuild.cs]: I can't follow as I read this as `AnnounceOnBuild.AnnounceOnlyToOwner`.

Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/Sound/AnnounceOnBuild.cs]: Then do not follow.

**Comment by Mailaender** [comment]
:+1:

**Comment by pchote** [comment]
Please make this use stances. It is silly to rewrite this later with upgrade rules when we can do it properly the first time around.

**Comment by GraionDilach** [comment]
Due to some circumstances, I definitely won't find time for that in the next week. Closing this then.

**Comment by GraionDilach** [comment]
It is also silly to scopecreep a supposed-to-be-5-minute-polish to ___replace every usage of Stance.Ally in the code with Stance.Owner and Stance.StrictAlly___ but after the CompletelyMissingThePoint PR, I got pissed off enough much that I actually started to work on that instead of some actual RL stuff I should care about.

**Comment by Phrohdoh** [comment]
I don't think asking a PR author to implement something in a consistent manner is scopecreep.

**Comment by SchismNavigator** [comment]
The whole point of this change is to enable the player to hear when their units are being built, and when they are fighting/winning.

I am not against it being expanded upon. But can we just get this basic functionality in first? It'll go a long way to adding more personality to units and personalising each faction.

**Comment by obrakmann** [comment]
@SchismNavigator: I don't see why those voice-overs should not be heard by the allies as well. The personality would still be limited to that faction. (Yeah, I know, that's what Generals & co. did, but idgaf.)

**Comment by GraionDilach** [comment]
Generals, Warcraft, Starcraft and co. uses the creation voiceovers as a replacement of the _Unit completed_ EVA message, to give a better oversight to the player which unit is done (since all these have individual queues and thereby one generic response is too vague). If a modder decided to use `AnnounceOnBuild` for this purpose, then allies shouldn't hear each other's units.

I don't understand `AnnounceOnKill` myself either, but I just imagine the same logic was applied to it ˙by them as well.

**Comment by Mailaender** [comment]
Being stubborn and blocking the PR caused an unnecessary fork: https://github.com/GraionDilach/OpenRA.Mods.AS/commit/0f4b0fe3c6ce8e8201762eee664e33251c2e35ee

**Comment by GraionDilach** [comment]
That's just a workaround. I'm ongoing with expanding the Stances flag, just a bit slow with updating and looking up all usecases and some RL issues which probably won't resolve themselves until early June as it stands atm.

**Comment by pchote** [comment]
> Being stubborn and blocking the PR caused an unnecessary fork: GraionDilach/OpenRA.Mods.AS@0f4b0fe

Downstream mods implementing their own custom logic in their own dll is a *good thing*.

**Comment by Mailaender** [comment]
It isn't own logic nor a good thing. That is copy paste of existing code with minor additions. Requiring this is an indicator of total failure.

**Comment by Mailaender** [comment]
Back to the topic. Do we really want a stance `Owner`?

**Comment by abcdefg30** [comment]
> Back to the topic. Do we really want a stance `Owner`?

Quite the late response... I'd say yes. ^^
Also: `[01:19:13] <penev> I'd say "yes" to that question`

**Comment by pchote** [comment]
I also agree that `Stance.Ally` is far to general to cover all our use cases.  Separating `Ally` and `Owner` is a good start, but may not be enough (or maybe just not the right name?) if we want to replace the `AffectsParent` flags used in warheads and external conditions.

**Comment by GraionDilach** [comment]
I actually had a working code back during those days where I kept Stance.Ally as-is and added Stance.StrictAlly and Stance.Owner (due to being lazy to write upgrade rules) and I only had regressions with shroud on first sight since shroud wasn't based on stances back those days.

AffectsParent should be kept unrelated to the stance setups, if the warhead is used to generate a wave explosion (missile with a rangelimit of 1 in bleed, InstantExplode projectile in AS) it would be expected that the firer wouldn't damage itself, but would still damage it's allied units owned by it's owner even - see the Firestorm Mobile EMP as an example, where friendly MEMPS do EMP each other down, but not themselves)

**Comment by pchote** [comment]
Ok

**Comment by GraionDilach** [comment]
I was wrong - Firestorm didn't have this logic (it used TypeImmune which means MEMPs are immune to themselves unrelated to owners), YR had it under the DamageSelf tag (used for the Nuke Silo, to prevent own silo damage but still damage enemy ones).

**Pull request closed without merging**